### PR TITLE
Experiment with generic EqualConstraint

### PR DESCRIPTION
**Pull request by CharliePoole**
**Experiment - Do Not Merge**

This branch is for the purpose of continuing a discussion Rob and I have been having about the use of generic constraints. We can continue here with the code in front of us and others are welcome to chime in.

In this branch, I have made EqualConstraint generic with a single type argument TExpected. That argument supplements the already existing TActual used in the generic ApplyTo method. Creating a class with two type args has been tried by both of us in the past and has run into issues, primarily because we don't know what the actual will be at the moment of instantiating the type.

I also added an overload to Assert.AreEqual using string types, so we would see the class being used with other than object types. Until we make more modifications to Assert, the constraint will only be used with objects.

Basically, going toward generic constraints has two aims: (1) providing more type-safety on the user side and (2) providing more efficiency by avoiding boxing of constants. So far, it doesn't look like we can get rid of the eventual boxing, but the type-safety is important enough by itself.

### CODE REVIEW

**Comment by rprouse** [comment]
@CharliePoole, I am going to bring our email conversation into this PR. I still prefer `AreEqual<T>` over `AreEqual<TActual, TExpected>`. I agree that we should allow some flexibility in the types passed in, but I think we can accomplish that. You mentioned examples of comparing a double to an int or an array to a list. These will work with an IEnumerable overload. 

For example, take a look at the following scratch code. All comparisons work where there is an automatic conversion and only one cast was needed between the double and the decimal. I think this gives us a good balance of constraining equality to types that can be compared but allowing flexible coding. It also has the advantage of allowing `Assert.AreEqual(null, value)`.

```C#
    class Program
    {
        private static void Main(string[] args)
        {
            int i = 2;
            decimal d = 2.0m;
            double db = 2.0;

            Assert.AreEqual(i, d);
            Assert.AreEqual(i, db);
            Assert.AreEqual(d, (decimal) db);

            string[] a = {"one", "two"};
            StringCollection sc = new StringCollection {"one", "two"};
            List<string> l = new List<string>{ "one", "two"};

            Assert.AreEqual(a, sc);
            Assert.AreEqual(a, l);
            Assert.AreEqual(sc, l);

            string s = "Hello world";

            Assert.AreEqual(null, s);

            Console.WriteLine("*** Press Enter to Exit ***");
            Console.ReadLine();
        }
    }

    public static class Assert
    {
        public static void AreEqual<T>(T expected, T actual)
        {
            Console.WriteLine("AreEqual<T> where T is {0}", typeof(T));
        }

        public static void AreEqual(IEnumerable expected, IEnumerable actual)
        {
            Console.WriteLine("AreEqual of IEnumerable");
        }
        
        // IEnumerable<T> overload is optional, but might give us more information
        public static void AreEqual<T>(IEnumerable<T> expected, IEnumerable<T> actual)
        {
            Console.WriteLine("AreEqual of IEnumerable<T> where T is {0}", typeof(T));
        }
    }
```

**Comment by CharliePoole** [comment]
I made a quick change to this branch and found 37 compiler errors in the framework tests. 

Comparisons of int or double with decimal, uint and byte require a cast. Generic arguments of test methods can't be compared for equality with any constant. Those aren't too bad.

Comparisons with nullable types get messed up in some cases, not others, dependent on order. That's bad.

Some tests compile using Assert.That but not Assert.AreEqual. That's unacceptable, but we can probably fix it if we end up making similar changes to Assert.That.

I'll work on this a little more and push an update.

**Comment by rprouse** [comment]
@CharliePoole, there are some good comments and history to this PR, but should we keep it open?

**Comment by CharliePoole** [comment]
No reason to leave it open.
